qiskit_experiments/calibration/experiments/spectroscopy.py

```diff
@@ -0,0 +1,180 @@
+
+        offset_guess = np.average(ydata)
+        amp_guess = np.max(ydata)
+        peak_idx = np.argmax(ydata)
+        x0_guess = xdata[peak_idx]
+        gamma_guesses = np.linspace(0, abs(xdata[-1] - xdata[0]), 10)
+
+        best_fit = None
+
+        for gamma_guess in gamma_guesses:
+            fit_result = curve_fit(
+                lambda x, a, g, x0, b: a * ( g**2 / ((x-x0)**2 + g**2)) + b,
+                xdata,
+                np.array(ydata),
+                np.array([amp_guess, gamma_guess, x0_guess, offset_guess]),
+                np.array(sigmas)
+            )
+
+            if not best_fit:
+                best_fit = fit_result
+            else:
+                if fit_result["reduced_chisq"] < best_fit["reduced_chisq"]:
+                    best_fit = fit_result
+
+        analysis_result = AnalysisResult(
+            {
+                "value": best_fit["popt"][2],
+                "stderr": best_fit["popt_err"][2],
+                "unit": unit,
+                "label": "Spectroscopy",
+                "fit": best_fit,
+                "quality": self._fit_quality(
+                    best_fit["popt"], best_fit["popt_err"], best_fit["reduced_chisq"]
+                ),
+            }
+        )
+
+        return analysis_result, None
+
+    @staticmethod
+    def _fit_quality(fit_out, fit_err, reduced_chisq):
+        """TODO"""
+        return ""
+
+
+class Spectroscopy(BaseExperiment):
+    """Class the runs spectroscopy by sweeping the qubit frequency."""
+
+    __analysis_class__ = SpectroscopyAnalysis
+
+    # Supported units for spectroscopy.
+    __units__ = {"Hz": 1.0, "kHz": 1.e3, "MHz": 1.e6, "GHz": 1.e9}
+
+    def __init__(self, qubit: int, frequency_shifts: Union[List[float], np.array], unit: Optional[str] = "Hz"):
+        """
+        A spectroscopy experiment run by shifting the frequency of the qubit.
+        The parameters of the GaussianSquare spectroscopy pulse are specified at run-time.
+        The spectroscopy pulse has the following parameters:
+        - amp: The amplitude of the pulse must be between 0 and 1, the default is 0.1.
+        - duration: The duration of the spectroscopy pulse in samples, the default is 1000 samples.
+        - sigma: The standard deviation of the pulse, the default is 5 x duration.
+        - width: The width of the flat-top in the pulse, the default is 0, i.e. a Gaussian.
+
+        Args:
+            qubit: The qubit on which to run spectroscopy.
+            frequency_shifts: The frequencies to scan in the experiment.
+            unit: unit of the frequencies: 'Hz', 'kHz', 'MHz', 'GHz'.
+
+        """
+        if len(frequency_shifts) < 3:
+            raise ValueError("Spectroscopy requires at least three frequencies.")
+
+        self._frequency_shifts = frequency_shifts
+
+        if unit not in self.__units__:
+            raise ValueError("Unsupported unit: {unit}.")
+
+        self._unit = unit
+        self._qubit = qubit
```

Comment:
This line should be removed. because the base class already has the `_qubits` data member. Then for example in the `circuits` method you can write `self._qubits[0]`.

Reply:
Indeed, see here fa1ee89